sklearn/metrics/pairwise_fast.pyx

```diff
@@ -41,28 +40,66 @@ def _chi2_kernel_fast(floating[:, :] X,
 
 def _sparse_manhattan(floating[::1] X_data, int[:] X_indices, int[:] X_indptr,
                       floating[::1] Y_data, int[:] Y_indices, int[:] Y_indptr,
-                      np.npy_intp n_features, double[:, ::1] D):
+                      double[:, ::1] D):
     """Pairwise L1 distances for CSR matrices.
 
     Usage:
-
     >>> D = np.zeros(X.shape[0], Y.shape[0])
-    >>> sparse_manhattan(X.data, X.indices, X.indptr,
-    ...                  Y.data, Y.indices, Y.indptr,
-    ...                  X.shape[1], D)
+    >>> _sparse_manhattan(X.data, X.indices, X.indptr,
+    ...                   Y.data, Y.indices, Y.indptr,
+    ...                   D)
     """
-    cdef double[::1] row = np.empty(n_features)
-    cdef np.npy_intp ix, iy, j
+    cdef np.npy_intp px, py, i, j, ix, iy
+    cdef double d = 0.0
 
-    with nogil:
-        for ix in range(D.shape[0]):
-            for iy in range(D.shape[1]):
-                # Simple strategy: densify current row of X, then subtract the
-                # corresponding row of Y.
-                memset(&row[0], 0, n_features * sizeof(double))
-                for j in range(X_indptr[ix], X_indptr[ix + 1]):
-                    row[X_indices[j]] = X_data[j]
-                for j in range(Y_indptr[iy], Y_indptr[iy + 1]):
-                    row[Y_indices[j]] -= Y_data[j]
+    cdef int m = D.shape[0]
+    cdef int n = D.shape[1]
+
+    # We scan the matrices row by row.
+    # Given row px in X and row py in Y, we find the positions (i and j
+    # respectively), in .indices where the indices for the two rows start.
+    # If the indices (ix and iy) are the same, the corresponding data values
+    # are processed and the cursors i and j are advanced.
+    # If not, the lowest index is considered. Its associated data value is
+    # processed and its cursor is advanced.
+    # We proceed like this until one of the cursors hits the end for its row.
+    # Then we process all remaining data values in the other row.
+
+    # Below the avoidance of inplace operators is intentional.
+    # When prange is used, the inplace operator has a special meaning, i.e. it
+    # signals a "reduction"
+
+    for px in prange(m, nogil=True):
```

Comment:
Have you experimented with chunking this computation? In other words, give each thread a group of `px`s to operate on and update `D` for that chunk.

Reply:
My understanding of the way OMP parallelizes for-loops is that the default schedule is "static" and that the static schedule does exactly what you just suggested, i.e. each thread is assigned a subrange of the same size (or of chunk_size, if you specify it).
I have not experimented with the other schedules, as this implementation already brought about a significant improvement.

Reply:
I was thinking about CPU cache misses. For example, if there are four threads and `px=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]` the work can be distributed as:

### Case One
```
[0, 1, 2], [3, 4, 5], [6, 7, 8], [9, 10, 11]
```

or

### Case Two
```
[0, 4, 8], [1, 5, 9], [2, 6, 10], [3, 7, 11]
```

In case one, each process will have access to chunks of memory close to each other. In case two, each process may go back to main memory to get data further apart.

Anyways this is a small nit.

Reply:
Yeah, as @ptocca mention I would also expect the default OMP scheduler to give each thread a contiguous chunk of memory to work with by default, unlike in the case two.

From https://cython.readthedocs.io/en/latest/src/userguide/parallelism.html#cython.parallel.prange
> static
> If a chunksize is provided, iterations are distributed to all threads ahead of time in blocks of the given chunksize. If no chunksize is given, the iteration space is divided into chunks that are approximately equal in size, and at most one chunk is assigned to each thread in advance.

Reply:
It does not mean it gives to each thread a contiguous chunk of memory. It just gives contiguous chunk of prange indices. It's up to us to make sure that contiguous range indexes correspond to contiguous chunks of memory. It seems to be the case here.